File: test/jsdom/armprogress.test.mjs
// The armed screen shows what each tier is doing, and never a blank wait (P06.S05, D16, D15, D34).
//
// ── What this is written against ─────────────────────────────────────────────
// `sessionStatus.diagnosis` is filled only once `bootstrapDone` is true, which is right for a
// VERDICT — a cause computed before the DHT has had its chance would accuse the wrong tier. But
// under ADR-011 nothing bootstraps until the local link has had its window, and where a browse has
// answered that hold is `lanFirstBudget`: thirty seconds. So the product was deliberately silent
// for the longest stretch of an arm, which is precisely the window D16 says must never be a blank
// spinner.
//
// ── Why the router's states are asserted separately ──────────────────────────
// D15's criterion has two halves and the second is the one that ships unexercised: the screen
// "discloses that a temporary router opening was requested and names the port; **when no mapping
// was obtained it says so rather than staying silent**". Silence, a refusal and an unroutable
// answer are three different next actions — D9's advice diverges to a VPN on the third — so a
// screen collapsing them gives one answer to three users who need different ones.
import test from 'node:test';
import assert from 'node:assert/strict';
import { boot } from './boot.mjs';

const PEER = { fingerprint: 'a'.repeat(64), label: 'Ada' };

let armed = false;
let progress = { link: 'watching', dht: 'holding' };
let until;

const { document: doc, settle } = await boot({
  routes: {
    '/api/peers': () => ({ self: 'b'.repeat(64), peers: [PEER] }),
    '/api/session/arm': () => { armed = true; return { armed: true, address: '127.0.0.1:8443' }; },
    '/api/session/status': () => ({ armed, address: '127.0.0.1:8443', progress, until }),
    '/api/session/disarm': () => { armed = false; return {}; },
  },
});

// The poll is scheduled 1200 ms after arming and this tier has no fake clock, so the wait is real
// — `waitdiagnosis.test.mjs`' own note, and the routes are re-read every poll, so a later case
// only needs the NEXT tick rather than another arm.
//
// **1800 ms and not 1400.** The poll interval is 1500 ms, so a tick shorter than one interval can
// return before the changed state has been fetched — measured: the `refused` case rendered the
// `silent` sentence, because the render it read was the previous poll's. A wait shorter than the
// thing it is waiting for is a flake that happens to pass most of the time.
const tick = async () => { await new Promise((r) => setTimeout(r, 1800)); await settle(); };

async function arm() {
  doc.getElementById('sessionRecvBtn').click();
  await settle();
  doc.getElementById('srvPeer').selectedIndex = 0;
  doc.getElementById('srvArmGo').click();
  await settle();
}

async function poll() {
  await tick();
  return doc.getElementById('srvWaitTiers');
}

test('the pre-bootstrap wait is described rather than left blank', async () => {
  await arm();
  const host = await poll();
  assert.ok(host, 'there is no tier list at all');
  assert.equal(host.hidden, false, 'the tier list is hidden while an arm is waiting');
  const text = host.textContent;
  assert.match(text, /local network/,
    'the link tier is not described. This is the window the diagnosis structurally cannot speak ' +
    'in, and it is the longest part of a LAN arm');
  assert.match(text, /Not using the internet yet/,
    'the DHT hold is not described. It is not a failure and not a delay to apologise for — it is ' +
    'the product deliberately not touching the public network until the link has had its chance');
});

test('the router opening names its port', async () => {
  progress = { link: 'found', dht: 'reaching', router: 'open', port: 41234 };
  const host = await poll();
  assert.match(host.textContent, /port 41234/,
    'D15 says the screen discloses that a temporary router opening was requested and NAMES THE ' +
    'PORT; a line saying only that one was opened is half the criterion');
});

test('no mapping says so, and says which kind of no', async () => {
  for (const [state, want] of [
    ['silent', /did not answer/],
    ['refused', /answered and declined/],
    ['unroutable', /cannot be reached from outside/],
  ]) {
    progress = { link: 'watching', router: state };
    const host = await poll();
    assert.match(host.textContent, want,
      `router=${state} is not described distinctly. Silence, a refusal and an unroutable answer ` +
      'are three different next actions — the third points at a VPN rather than a port-forward — ' +
      'and one sentence for all three gives one answer to three users');
  }
  // And they are DIFFERENT sentences, not one string that happens to match three patterns.
  const seen = new Set();
  for (const state of ['silent', 'refused', 'unroutable']) {
    progress = { link: 'watching', router: state };
    seen.add((await poll()).textContent);
  }
  assert.equal(seen.size, 3, 'the three router failures render the same text');
});

test('no tier line is a countdown', async () => {
  until = '2026-10-01T12:00:00Z';
  progress = { link: 'found', dht: 'reaching', router: 'open', port: 41234 };
  const host = await poll();
  // D16's amendment: only the ceremony deadline appears in human units, and neither the connect
  // deadline nor the exchange deadline appears as a countdown. A tier line showing seconds left
  // would be exactly that — and it would invite a user to watch a number instead of reading the
  // router line, which is the only one they can act on.
  assert.doesNotMatch(host.textContent, /\b\d+\s*(s|sec|second|m|min|minute)s?\b/i,
    `a tier line reads like a countdown: ${JSON.stringify(host.textContent)}`);
});

test('the tier list clears when the arm stops reporting', async () => {
  progress = { link: 'watching' };
  assert.equal((await poll()).hidden, false, 'setup: nothing was shown to clear');
  progress = undefined;
  const host = await poll();
  assert.equal(host.hidden, true,
    'the ladder is still on screen after the arm stopped. Progress is LIVE state, not a sticky ' +
    'record — leaving it up describes a wait nobody is doing');
});

// **Disarmed at the end, and it is not tidiness.** `pollRecv` reschedules itself every 1.5 s for as
// long as the arm is up, and node's test runner waits for a quiet event loop — a file that leaves
// one armed keeps a timer alive and the whole suite hangs behind it. Measured: the run went from
// ten seconds to the harness's 900-second ceiling with this missing.
test('the arm is put away, so this file leaves no timer behind', async () => {
  doc.getElementById('srvDisarm')?.click();
  await settle();
  armed = false;
  await tick();
  assert.equal(armed, false, 'the arm is still up');
});

File: internal/server/armprogress_test.go
package server

import (
	"testing"
	"time"
)

// P06.S05 — what each tier of the connection ladder is doing, while it is doing it.

// TestTheArmProgressIsNotGatedOnTheBootstrap is the criterion's own defect.
//
// **The blank spinner was not hypothetical — it was the shipped state, and its cause was a gate.**
// `sessionStatus` fills `Diagnosis` only `if cer != nil && !inSession && cer.bootstrapDone.Load()`.
// That is right for a VERDICT: a cause computed before the DHT has had its chance would accuse the
// wrong tier. But under ADR-011 nothing bootstraps until the local link has had its window, and
// where a browse has answered that hold is `lanFirstBudget` — thirty seconds — so the product was
// deliberately silent for the longest stretch of an arm. D16 says that window must never be a
// blank spinner.
//
// So this drives the state the diagnosis cannot speak in: watching the link, bootstrap NOT done.
func TestTheArmProgressIsNotGatedOnTheBootstrap(t *testing.T) {
	cer := &ceremonyID{}
	cer.watchingLink(time.Now())

	// SETUP: the bootstrap really has not run, or this is not the window under test and the
	// assertion below would pass in a state the diagnosis could already describe.
	if cer.bootstrapDone.Load() {
		t.Fatal("setup: the bootstrap is already done, so this is not the pre-bootstrap window " +
			"the criterion is about")
	}

	p := cer.armProgressOf(time.Now())
	if p == nil {
		t.Fatal("an arm watching the link reports NOTHING before the bootstrap. That is the blank " +
			"spinner D16 forbids, and it is the longest part of a LAN arm by ADR-011's design")
	}
	if p.Link != "watching" {
		t.Errorf("link=%q, want watching", p.Link)
	}
	if p.DHT != "holding" {
		t.Errorf("dht=%q, want holding — the wait is not a failure and not a delay to apologise "+
			"for: it is the product deliberately not touching the public network until the link "+
			"has had its chance, and it is the state the screen has never had a word for", p.DHT)
	}
}

// TestTheLinkTierReportsFoundOnlyOnItsOwnPeersSighting.
//
// ADR-011's hold renews on evidence — *"every resolved sighting of its own expected peer renews the
// hold"* — and `linkSeenAt` is set by that path alone. A screen reporting "found" for any
// announcement would tell a user the ceremony is progressing when nothing of theirs has been seen.
func TestTheLinkTierReportsFoundOnlyOnItsOwnPeersSighting(t *testing.T) {
	cer := &ceremonyID{}
	cer.watchingLink(time.Now())
	if got := cer.armProgressOf(time.Now()).Link; got != "watching" {
		t.Fatalf("setup: link=%q before any sighting, want watching", got)
	}
	cer.noteLinkSighting(time.Now())
	if got := cer.armProgressOf(time.Now()).Link; got != "found" {
		t.Errorf("after a sighting of its own expected peer link=%q, want found", got)
	}
}

// TestTheRouterTierKeepsItsThreeFailuresApart is D15's absence clause, and the arm most likely to
// ship unexercised.
//
// **Silence, refusal and unroutable are three different next actions**, which is why the state is
// not a boolean. Silence may mean there is no gateway to ask. A refusal means *"the router is the
// user's, is reachable, and said no"* (`mapRefused`'s own words). An unroutable answer means a
// second layer of NAT, and D9's advice diverges there to a VPN rather than a port-forward. A screen
// that said "no mapping" for all three would give one answer to three users who need different ones.
func TestTheRouterTierKeepsItsThreeFailuresApart(t *testing.T) {
	// Never asked: no mapper at all. Distinct from asking and hearing nothing.
	cer := &ceremonyID{}
	cer.watchingLink(time.Now())
	if got := cer.armProgressOf(time.Now()).Router; got != "" {
		t.Errorf("an arm that never asked for a mapping reports router=%q, want empty", got)
	}

	// Asked, nothing answered.
	silent := &ceremonyID{}
	silent.watchingLink(time.Now())
	silent.portMap = &portMapper{}
	if got := silent.armProgressOf(time.Now()).Router; got != "silent" {
		t.Errorf("router=%q for a mapper that obtained nothing, want silent", got)
	}

	// A gateway answered and said no.
	refused := &ceremonyID{}
	refused.watchingLink(time.Now())
	refused.portMap = &portMapper{}
	refused.mapRefused = true
	if got := refused.armProgressOf(time.Now()).Router; got != "refused" {
		t.Errorf("router=%q when a gateway answered with a refusal, want refused — the opposite "+
			"fact from silence, and D9 tells those two users different things", got)
	}

	// An answer that cannot be published.
	unroutable := &ceremonyID{}
	unroutable.watchingLink(time.Now())
	unroutable.portMap = &portMapper{}
	unroutable.mapUnroutable = true
	if got := unroutable.armProgressOf(time.Now()).Router; got != "unroutable" {
		t.Errorf("router=%q for an unpublishable answer, want unroutable — D9's advice diverges "+
			"here to a VPN rather than a port-forward", got)
	}
}

// TestAnArmWithNothingToSaySaysNothing.
//
// A progress view of four empty strings is a ladder with no rungs, and a client would render it as
// one. The manual and plain co-sign paths have no ladder to report on at all.
func TestAnArmWithNothingToSaySaysNothing(t *testing.T) {
	if p := (*ceremonyID)(nil).armProgressOf(time.Now()); p != nil {
		t.Errorf("a nil ceremony reports progress %+v", p)
	}
	if p := (&ceremonyID{}).armProgressOf(time.Now()); p != nil {
		t.Errorf("an arm that is watching nothing, has not bootstrapped and never asked for a "+
			"mapping reports %+v — four empty strings a client draws as an empty ladder", p)
	}
}

// TestTheStatusPublishesProgressBeforeTheBootstrap is where the criterion's defect actually lived.
//
// **The unit tests above grade `armProgressOf`; this grades the GATE.** `sessionStatus` fills
// `Diagnosis` only when `cer.bootstrapDone.Load()`, and the whole finding of this slice is that
// putting progress behind the same condition reproduces the blank spinner exactly — the tier view
// would arrive at the moment the diagnosis could already speak, which is after the window it exists
// for. So this asserts the status carries progress while the bootstrap has NOT happened.
//
// A test that only drove `armProgressOf` would stay green against a status that never published it.
func TestTheStatusPublishesProgressBeforeTheBootstrap(t *testing.T) {
	var se session
	cer := &ceremonyID{}
	cer.watchingLink(time.Now())
	se.arms[armInteractive] = &arm{cer: cer, addr: "127.0.0.1:9999"}

	// SETUP, both halves. The bootstrap has not run, and the DIAGNOSIS is therefore silent — which
	// is the state this whole slice is about. Without the second check this test could pass in a
	// window where the diagnosis was already speaking and prove nothing about the gate.
	if cer.bootstrapDone.Load() {
		t.Fatal("setup: the bootstrap is done, so this is not the window under test")
	}
	st := se.status()
	if st.Diagnosis != nil {
		t.Fatalf("setup: the diagnosis is already speaking (%+v), so a blank screen was never "+
			"the state here and the assertion below is not about the gate", st.Diagnosis)
	}

	if st.Progress == nil {
		t.Fatal("the status carries NO progress while the bootstrap has not run. That is the " +
			"blank spinner: the diagnosis is gated on bootstrapDone so it cannot speak yet, and " +
			"under ADR-011 the bootstrap waits for the local link — thirty seconds on a LAN by " +
			"lanFirstBudget. D16 says that window must never be blank, and it is the window")
	}
	if st.Progress.Link != "watching" || st.Progress.DHT != "holding" {
		t.Errorf("progress is %+v, want link=watching dht=holding", st.Progress)
	}
}
